serverless/cmd/feeder/github/main.go

```diff
@@ -0,0 +1,561 @@
+}
+
+func setupWitnessRepo(ctx context.Context, opts *options) (*git.Repository, error) {
+	// Clone the fork of the log, so we can go on to make a PR against it.
+	//   git clone -o origin "https://${GIT_USERNAME}:${FEEDER_GITHUB_TOKEN}@github.com/${fork_repo}.git" "${temp}"
+	cloneOpts := &git.CloneOptions{
+		URL: fmt.Sprintf("https://%v:%v@github.com/%s.git", opts.gitUsername, opts.githubAuthToken, opts.witnessRepo),
+	}
+
+	var forkRepo *git.Repository
+	var err error
+	dest := opts.witnessClonePath
+	if opts.witnessClonePath == "" {
+		forkRepo, err = git.Clone(memory.NewStorage(), memfs.New(), cloneOpts)
+		dest = "memory"
+	} else {
+		forkRepo, err = git.PlainClone(opts.witnessClonePath, false, cloneOpts)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to clone fork repo %q: %v", cloneOpts.URL, err)
+	}
+	glog.V(1).Infof("Cloned %q into %q", cloneOpts.URL, dest)
+
+	// Create a remote -> logOwnerRepo
+	//  git remote add upstream "https://github.com/${log_repo}.git"
+	logURL := fmt.Sprintf("https://github.com/%v.git", opts.logRepo)
+	logRemote, err := forkRepo.CreateRemote(&config.RemoteConfig{
+		Name: "upstream",
+		URLs: []string{logURL},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to add remote %q to fork repo: %v", logURL, err)
+	}
+	glog.V(1).Infof("Added remote upstream->%q for %q:\n%v", logURL, opts.witnessRepo, logRemote)
+
+	//  git fetch --all
+	if err = forkRepo.FetchContext(ctx, &git.FetchOptions{}); err != nil && err != git.NoErrAlreadyUpToDate {
+		return nil, fmt.Errorf("failed to fetch repo %q: %v", *opts.witnessRepo, err)
+	}
+
+	return forkRepo, nil
+}
+
+func authWithGithub(ctx context.Context, ghToken string) (*github.Client, error) {
+	ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: ghToken})
+	tc := oauth2.NewClient(ctx, ts)
+	return github.NewClient(tc), nil
+}
+
+func pullAndGetRepoHead(r *git.Repository) (*plumbing.Reference, *git.Worktree, error) {
+	// Figure out where the worktree is for repo 'r'.
+	wt, err := r.Worktree()
+	if err != nil {
+		return nil, nil, fmt.Errorf("workTree(%v) err: %v", r, err)
+	}
+	// Pull the latest commits from remote 'origin'.
+	if err := wt.Pull(&git.PullOptions{RemoteName: "origin"}); err != nil && err != git.NoErrAlreadyUpToDate {
+		return nil, nil, fmt.Errorf("git pull %v/origin err: %v", r, err)
+	}
+	// Get the master HEAD - we'll need this to branch from later on.
+	headRef, err := r.Head()
+	if err != nil {
+		return nil, nil, fmt.Errorf("reading %v HEAD ref err: %v", r, err)
+	}
+	// Print the latest commit that was just pulled
+	commit, err := r.CommitObject(headRef.Hash())
+	if err != nil {
+		return nil, nil, fmt.Errorf("reading %v HEAD commit err: %v", r, err)
+	}
+	glog.V(1).Infof("Reading %v: HEAD commit:\n%v", r, commit)
+	return headRef, wt, nil
+}
+
+// feedOnce performs a one-shot "feed to witness and create PR" operation.
+func feedOnce(ctx context.Context, opts *options, forkRepo *git.Repository, ghCli *github.Client) error {
```

Comment:
This would probably be better moved further up the file. I'd stick it near the bottom of `main` as it's effectively the main loop of the function there.

Reply:
👍  done